### DISCUSSION TITLE
Discrete nucleation overlap

### DISCUSSION THREAD

**Discussion by achnr** [idaholab/moose#22182] [Q&A Modules: Phase field] [answered]
Hi,

I'm using Discrete Nucleation for a recrystallization problem. I noticed that new nuclei are allowed to overlap with previous ones. This is problematic because it unnecessarily creates (extra) new grains which requires more order parameters and refines the mesh even more. 
Is there a way to prevent this overlap?

Thanks,
Amani

**Comment by GiudGiud** [comment, marked as answer]
@laagesen @permcody any ideas on this?

**Reply by permcody** [reply]
@dschwen may have a suggestion as well. The GrainTracker can handle nucleation, but the nucleation objects are responsible for avoiding overlap and I believe this should be working properly. The problem may not be setup correctly.

**Reply by laagesen** [reply]
Hi Amani - The DiscreteNucleationInserter requires a material property that gives the probability of nucleation- in a past nucleation problem I was working on where voids were nucleating in a solid matrix, I made the material property include a conditional that set the nucleation probability to zero unless the matrix order parameter was larger than a certain value (0.98) close to the equilibrium value (1), to avoid nuclei forming on top of existing ones. This was on top of the probability dependence on classical nucleation theory, which on its own wasn't enough to keep nuclei from forming close to existing ones. I'm not sure if that same approach will work for a recrystallization problem but maybe it can be adapted...

**Reply by achnr** [reply]
Thank you for your answers. I had essentially used the same approach as @laagesen, however the center of a new nucleus could still be in the vicinity of a pre-existing nucleus, which creates an overlap between parts of both nuclei. With recrystallization, the number of new nuclei can be significant enough to have many overlaps in the domain. 

The other issue is that within one time step, new nuclei overlap with each other as the probability is not updated until the next time step. It seems that the nucleation algorithm could be improved in that regard.

**Reply by FHilty** [reply]
Hi Everyone, I am also working on a recrystallization model and ran into similar overlap issues. I have spent some time fixing the problem of overlap within a single time step by modifying `DescreteNucleationInserter`.  In that user object each mpi rank evaluates its piece of the domain and generates a list of qp for new nuclei centers. These local lists are then collected together, after which it is easy to check if any of the new nuclei overlap and remove one when it does.

The problem is this method can skew the nucleation rate. The image below shows an extreme case with a high nucleation rate and large nuclei (left is the original code and right is with my changes to remove overlap), where half of the nuclei are lost. This effect should be minimized with smaller nuclei and slower rates, such that fewer are eliminated at this step, but it is still a potential issue.

There does not seem to be a straightforward way to correct nucleation rate problem in this case due to the nuclei being generated in each mpi rank, and full list only existing to check afterward. Which is possibly why no routine to prevent the overlap has been implemented already. I was thinking about having the fraction of nuclei that were eliminated printed to screen so the user could decide if it was a significant nucleation rate reduction or not, but i would love to hear ideas about how to properly fix the root problems.

![NucleationOverlap](https://user-images.githubusercontent.com/14181200/194672434-30dfa669-2883-4638-9c3d-8fd6b763cfa5.png)
